src/triage/experiments/base.py

```diff
@@ -84,7 +84,7 @@ def __init__(
         self._check_config_version(config)
         self.config = config
 
-        if isinstance(db_engine, Engine):
```

Comment:
One thing that came to mind is that maybe this check should be in multicore. There shouldn't be any harm in converting this for SingleThreaded, but just for the sake of the check being closer to the reason for the check (e.g. multiprocess code). What do you think?

Reply:
If we do so, I tried to see if there's an even better check, for instance trying to get the pickler directly from the multiprocessing library we're using....but I don't see that exposed at all. We could go full duck-type check and actually fork a process with the engine and see if it works.

Reply:
Yeah I isolated the multiprocessing pickler somewhere at some point ... (And, I like the idea of "duck"-typing.)

I think this would be it:

```
from multiprocessing.reduction import ForkingPickler

ForkingPickler.dumps(…)
```

And, regardless, I agree that this logic should be closer to the multiprocessing code, (in `multicore`).

Reply:
That said, I don't love the extra, eager pickling -- bear in mind, rather than duck-typing, we could simply say:

    if isinstance(db_engine, SerializableDbEngine):
        …
    else:
        …

And, eagerly spawning a test fork really seems like overkill.

In fact, **if** it's possible to reliably detect a child process / multiprocessing issue due to serialization, then why check eagerly at all? Could we just let the experiment go, but detect the serialization error and exit (propagating the error)?

That might be nicer than the route of warning and wrapping. (Especially considering that that wrapping code only carries through the DB URL, and the fact that your engine has been replaced with ours *can* be overlooked.)

Reply:
There's value in raising the error immediately after the user starts running and not minutes (if not hours) later when we first encounter the problem, because with the runtimes of experiments people are likely to start them, let them go, and come back later.

Reply:
I'm curious your problem with duck-typing, especially since in a prior comment you said you liked the idea. Is there a performance concern with eagerly pickling something so small as an engine?

Reply:
Yep, I just meant that if we can ensure that the (multiprocessing) procedure raises pickling errors sufficiently early, then this could be better than picking out the engine for pickling that eagerly.

This is only very slightly a performance consideration – (that's likely *not* a significant implication). But, also, because detecting a troublesome engine may be finicky, and because there could be other serialization issues beyond just the engine.

Duck-typing is good; though, it takes various forms, and often isn't "let's just _do the thing_ early to ensure we can do it again later" – though that's not an unreasonable procedure. (Often it's rather "does it 'quack' like a duck," which means checking for indicators of its functionality; but, which I don't know that we _can_ check for, here.)

And, depending on the situation, it is at least as good, and less finicky, to simply "try first, ask questions (that is catch exceptions) later." And depending on the multiprocessing tool I use, this even seems possible – though, perhaps it's not feasible, here.

Reply:
Anyway, now having been reduced to tears by the `multiprocessing` and `pebble` modules … :man_shrugging: :smile: 

I can get `multiprocessing.Pool.map` to raise eager pickling errors -- and it makes sense to catch and propagate those, through the parent process, to the user.

However, that's not what I get with `pebble.ProcessPool.map`; and, I don't know that there's an profitable path to get that, though maybe you have more insight.

And, if we're rather sure that the other components of our multiprocessing task payloads -- other than the engine that is -- "come from us" and should be serializable, then I don't see great value in checking the payloads at the lower level.

So if that's all the case, we are back here, checking the engine :smile:

Reply:
Referring to the duck typing discussion: I get that sometimes we do want to wait until we need to do the thing before we test it out, but I think in triage experiments our priorities are a bit different. 

Referring to the eager process forking: It's a little bit hard to really test all of the process forking in the context that it will be done in the end: A bunch of different components have engines, as well as other arguments that may or may not be serializable. To go that far we would essentially need to do a dry run, or sample run, of the entire experiment with a tiny subset of data.

This 'sample data run' is something we do want to work towards (per the discussion after my presentation on testing a month ago), but it does seem a bit overboard for a bugfix like this.